core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -734,9 +981,52 @@ impl<Block: BlockT> network_gossip::Validator<Block> for GossipValidator<Block>
 	fn validate(&self, context: &mut dyn ValidatorContext<Block>, who: &PeerId, data: &[u8])
 		-> network_gossip::ValidationResult<Block::Hash>
 	{
-		let (action, broadcast_topics) = self.do_validate(who, data);
+		let (action, broadcast_topics, peer_reply) = self.do_validate(who, data);
+
+		let catch_up = {
```

Comment:
Why do this within `fn validate`? If anything, I'd expect this logic to live within the processing logic for neighbor packets.

Reply:
Indeed this was a bit messy.